### DISCUSSION TITLE
New module NNG proxy is finished.but requires docs & demo & configuration files

### DISCUSSION THREAD

**Discussion by JaylinYu** [nanomq/nanomq#646] [Ideas]
now NanoMQ is able to proxy msg between brokerless and broker networks. If anyone is interested in having a try on this feature, I would be happy to provide support.

**Comment by brugr9** [comment]


**Reply by JaylinYu** [reply]
Hi Roland, 
Thanks for pointing out my mistakes, there is a simple arg parsing error, now it has been fixed!

**Reply by xinyi-xs** [reply]
Hi @brugr9 ,  the document first line and third line both for demo to show how to use this tool. First line shows NNG/SP listen port, if you have done it (ignore it) and the third line is just represents all your mqtt dataflow. Based on your description，'Any Data Producer', you can use this command (subscribe '#' topic to aggregate all topic) to get all topic data flow and proxy to port 5555.
```shell
nanomq_cli nngproxy pub0 --mqtt_url "mqtt-tcp://localhost:1883" --dial "tcp://127.0.0.1:5555" -t '#' --qos 0
```
before run this command, you should launch NanoMQ listen on port 1883 and NNG/SP listen on port 5555

**Reply by xinyi-xs** [reply]
Hi @brugr9 , can you tell me in detail about your environment and version, I can't reproduce your problem locally, or you can try the latest version to see if the above problem still occurs.

**Reply by JaylinYu** [reply]
Hi brugr9, currently nngproxy does only proxying msg from an nng pub/sub client to an MQTT broker as its name says. So it means you need to have both nng client and broker ready then start this gateway.

And I noticed that you also got "nng_dialer_create:Not supported" output, which means the nng lib shipped with docker image you installed is not nanonng or somehow does not support MQTT.

**Reply by JaylinYu** [reply]
The proxy is a bidirectional implementation, just replace pub0 with sub0. Gonna write a blog/tutorial for this.
I am not sure what happens with your nng lib. But multiple nng installation in your env is highly suspected.
